internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,570 @@
+//
+// Return Value:
+//   - nil: indicates that the authorization check passed and the operation is permitted.
+//   - non-nil error: indicates that an error occurred during the permission evaluation process
+//     (for example, a failure decoding the manifest or other internal issues). If the evaluation
+//     completes successfully but identifies missing rules, then a nil error is returned along with
+//     the list (or slice) of missing rules. Note that in some cases the error may encapsulate multiple
+//     evaluation failures
+func (a *rbacPreAuthorizer) PreAuthorize(ctx context.Context, manifestManager user.Info, manifestReader io.Reader) ([]ScopedPolicyRules, error) {
+	var allMissingPolicyRules = []ScopedPolicyRules{}
+	dm, err := a.decodeManifest(manifestReader)
+	if err != nil {
+		return nil, err
+	}
+	attributesRecords := dm.asAuthorizationAttributesRecordsForUser(manifestManager)
+
+	var preAuthEvaluationErrors []error
+	missingRules, err := a.authorizeAttributesRecords(ctx, attributesRecords)
+	if err != nil {
+		preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+	}
+
+	ec := escalationChecker{
+		authorizer:        a.authorizer,
+		ruleResolver:      a.ruleResolver,
+		extraClusterRoles: dm.clusterRoles,
+		extraRoles:        dm.roles,
+	}
+
+	for _, obj := range dm.rbacObjects() {
+		if err := ec.checkEscalation(ctx, manifestManager, obj); err != nil {
+			missingEscalationRules, namespace := parseEscalationErrorForMissingRules(err)
+			// Check if we already have these escalation PolicyRules, if so don't append
+			for i, rule := range missingEscalationRules {
+				previousRule := missingRules[namespace][len(missingRules[namespace])-len(missingEscalationRules)+i]
+				if !arePolicyRulesEqual(previousRule, rule) {
+					missingRules[namespace] = append(missingRules[namespace], missingEscalationRules...)
+				}
+			}
+			preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+		}
+	}
+
+	for ns, nsMissingRules := range missingRules {
+		// NOTE: Although CompactRules is defined to return an error, its current implementation
+		// never produces a non-nil error. This is because all operations within the function are
+		// designed to succeed under current conditions. In the future, if more complex rule validations
+		// are introduced, this behavior may change and proper error handling will be required.
+		if compactMissingRules, err := validation.CompactRules(nsMissingRules); err == nil {
+			missingRules[ns] = compactMissingRules
+		}
+		sortableRules := rbacv1helpers.SortableRuleSlice(missingRules[ns])
+		sort.Sort(sortableRules)
+		allMissingPolicyRules = append(allMissingPolicyRules, ScopedPolicyRules{Namespace: ns, MissingRules: sortableRules})
+	}
+
+	// sort allMissingPolicyRules alphabetically by namespace
+	sort.Slice(allMissingPolicyRules, func(i, j int) bool {
+		return allMissingPolicyRules[i].Namespace < allMissingPolicyRules[j].Namespace
+	})
```

Comment:
Nit: the new `slices` package in the standard library has `slices.SortFunc` which is a generics-based sort implementation that makes this a bit more ergonomic (the less func allows direct comparison of the two objects rather than requiring the index lookup).

Reply:
Done